Greenmask V0.1.13 SIGSEGV

Hi!

While using Greenmask for the first time, I've encountered a segfault. Using the `dump` option while testing on a single table in my database. I've verified connection with `psql` following config:
```
dump:
  pg_dump_options:
    dbname: "host=obfuscated-amazon-address user=postgres dbname=db_name"
    jobs: 10
    table: "tablename"
storage:
  type: "directory"
  directory:
    path: "/home/ssm-user/tmp"
```

The segfault output is:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xd3ddb1]

goroutine 151 [running]:
github.com/jackc/pgx/v5.(*LargeObject).Close(...)
	/home/runner/go/pkg/mod/github.com/jackc/pgx/v5@v5.5.5/large_objects.go:156
github.com/greenmaskio/greenmask/internal/db/postgres/dumpers.(*BlobsDumper).Execute.func2.2()
	/home/runner/work/greenmask/greenmask/internal/db/postgres/dumpers/large_object.go:95 +0x31
github.com/greenmaskio/greenmask/internal/db/postgres/dumpers.(*BlobsDumper).Execute.func2()
	/home/runner/work/greenmask/greenmask/internal/db/postgres/dumpers/large_object.go:100 +0x1f1
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/home/runner/go/pkg/mod/golang.org/x/sync@v0.6.0/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 65
	/home/runner/go/pkg/mod/golang.org/x/sync@v0.6.0/errgroup/errgroup.go:75 +0x96
```


For completeness; The `uname -r` output gives: `6.1.79-99.167.amzn2023.x86_64`. I've used `greenmask-linux-amd64.tar.gz` and the following libraries `postgresql15 postgresql15-contrib`. The database version is Postgres 15 as well.

On a different note, I've also found that `greenmask dump --data-only` gives a schema collision error : `"pg_dump: error: options -s/--schema-only and -a/--data-only cannot be used together"`

--- a/internal/db/postgres/dumpers/large_object.go
+++ b/internal/db/postgres/dumpers/large_object.go
@@ -56,67 +56,10 @@ func (lod *BlobsDumper) Execute(ctx context.Context, tx pgx.Tx, st storages.Stor
 		w, r := countwriter.NewGzipPipe()
 
 		// Writing goroutine
-		eg.Go(
-			func() error {
-				defer func() {
-					log.Debug().
-						Uint32("oid", uint32(lo.Oid)).
-						Msg("closing reader")
-					if err := r.Close(); err != nil {
-						log.Warn().
-							Err(err).
-							Uint32("oid", uint32(lo.Oid)).
-							Msg("error closing LargeObject reader")
-					}
-				}()
-				err := st.PutObject(gtx, fmt.Sprintf("blob_%d.dat.gz", lo.Oid), r)
-				if err != nil {
-					return fmt.Errorf("cannot write large object %d object: %w", lo.Oid, err)
-				}
-				return nil
-			},
-		)
+		eg.Go(largeObjectWriter(gtx, st, lo, r))
 
 		// Dumping goroutine
-		eg.Go(
-			func() error {
-				defer func() {
-					log.Debug().
-						Uint32("oid", uint32(lo.Oid)).
-						Msg("closing writer")
-					if err := w.Close(); err != nil {
-						log.Warn().Err(err).Msg("error closing Blobs writer")
-					}
-				}()
-				buf := make([]byte, loBufSize)
-				los := tx.LargeObjects()
-				lo, err := los.Open(ctx, uint32(lo.Oid), pgx.LargeObjectModeRead)
-				defer func() {
-					if err := lo.Close(); err != nil {
-						log.Warn().Err(err).Msg("error closing large object")
-					}
-				}()
-				if err != nil {
-					return fmt.Errorf("error dumping large object %d: %w", lo, err)
-				}
-				var done bool
-				for !done {
-					size, err := lo.Read(buf)
-					if err != nil {
-						if errors.Is(err, io.EOF) {
-							buf = buf[:size]
-							done = true
-						} else {
-							return fmt.Errorf("error reading large object %d: %w", lo, err)
-						}
-					}
-					if _, err = w.Write(buf); err != nil {
-						return fmt.Errorf("error writing large object %d into storage: %w", lo, err)
-					}
-				}
-				return nil
-			},
-		)
+		eg.Go(largeObjectDumper(gtx, lo, w, tx))
 
 		if err := eg.Wait(); err != nil {
 			return nil, err
@@ -143,6 +86,70 @@ func (lod *BlobsDumper) Execute(ctx context.Context, tx pgx.Tx, st storages.Stor
 	}
 
 	return lod.Blobs, nil
+}
+
+func largeObjectWriter(ctx context.Context, st storages.Storager, lo *entries.LargeObject, r countwriter.CountReadCloser) func() error {
+	return func() error {
+		defer func() {
+			log.Debug().
+				Uint32("oid", uint32(lo.Oid)).
+				Msg("closing reader")
+			if err := r.Close(); err != nil {
+				log.Warn().
+					Err(err).
+					Uint32("oid", uint32(lo.Oid)).
+					Msg("error closing LargeObject reader")
+			}
+		}()
+		err := st.PutObject(ctx, fmt.Sprintf("blob_%d.dat.gz", lo.Oid), r)
+		if err != nil {
+			return fmt.Errorf("cannot write large object %d object: %w", lo.Oid, err)
+		}
+		return nil
+	}
+}
+
+func largeObjectDumper(ctx context.Context, lo *entries.LargeObject, w countwriter.CountWriteCloser, tx pgx.Tx) func() error {
+	return func() error {
+		defer func() {
+			log.Debug().
+				Uint32("oid", uint32(lo.Oid)).
+				Msg("closing writer")
+			if err := w.Close(); err != nil {
+				log.Warn().Err(err).Msg("error closing Blobs writer")
+			}
+		}()
+		buf := make([]byte, loBufSize)
+		los := tx.LargeObjects()
+		loObj, err := los.Open(ctx, uint32(lo.Oid), pgx.LargeObjectModeRead)
+		if err != nil {
+			return fmt.Errorf("error opening large object %d: %w", lo.Oid, err)
+		}
+		defer func() {
+			if err := loObj.Close(); err != nil {
+				log.Warn().Err(err).Msg("error closing large object")
+			}
+		}()
+		if err != nil {
+			return fmt.Errorf("error dumping large object %d: %w", lo.Oid, err)
+		}
+		var done bool
+		for !done {
+			size, err := loObj.Read(buf)
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					buf = buf[:size]
+					done = true
+				} else {
+					return fmt.Errorf("error reading large object %d: %w", lo.Oid, err)
+				}
+			}
+			if _, err = w.Write(buf); err != nil {
+				return fmt.Errorf("error writing large object %d into storage: %w", lo.Oid, err)
+			}
+		}
+		return nil
+	}
 }
 
 func (lod *BlobsDumper) DebugInfo() string {

--- a/internal/db/postgres/dumpers/large_object.go
+++ b/internal/db/postgres/dumpers/large_object.go
@@ -74,6 +74,16 @@ func (lod *BlobsDumper) Execute(ctx context.Context, tx pgx.Tx, st storages.Stor
 	}
 
 	// Writing blobs.toc
+	if err := lod.generateBlobsToc(ctx, st); err != nil {
+		return nil, fmt.Errorf("cannot write large object blobs.toc: %w", err)
+	}
+
+	return lod.Blobs, nil
+}
+
+func (lod *BlobsDumper) generateBlobsToc(ctx context.Context, st storages.Storager) error {
+	log.Debug().Msg("writing blobs.toc")
+	// Writing blobs.toc
 	blobsTocBuf := bytes.NewBuffer(nil)
 
 	for _, lo := range lod.Blobs.LargeObjects {
@@ -82,10 +92,9 @@ func (lod *BlobsDumper) Execute(ctx context.Context, tx pgx.Tx, st storages.Stor
 
 	err := st.PutObject(ctx, "blobs.toc", blobsTocBuf)
 	if err != nil {
-		return nil, fmt.Errorf("cannot write large object blobs.toc: %w", err)
+		return fmt.Errorf("cannot write large object blobs.toc: %w", err)
 	}
-
-	return lod.Blobs, nil
+	return nil
 }
 
 func largeObjectWriter(ctx context.Context, st storages.Storager, lo *entries.LargeObject, r countwriter.CountReadCloser) func() error {

